<a href="https://colab.research.google.com/github/Abdurrasyiid/Dasar-Machine-Learning/blob/master/Modul%201%20Data%20Introduction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Data Processing

Data processing adalah tahap paling awal dalam setiap proyek ML. Pada tahap ini data akan diambil dari sumber tertentu, dimasukkan pada suatu environment, dan diproses agar bisa diolah oleh model machine learning.



> Pandas adalah sebuah library open source yang dipakai untuk menganalisis dan memanipulasi data. Pandas dibangun menggunakan bahasa pemrograman Python yang menawarkan struktur data dan operasi untuk manipulasi tabel numerik dan time series. Tabel numerik adalah tabel yang berisi bilangan numerik, dan Tabel time series adalah tabel yang berubah seiring waktu, misalnya tabel yang memuat perubahan nilai pasar saham untuk setiap menitnya.





> Pandas merupakan salah satu alternatif library yang sering digunakan untuk proses data wrangling. Pandas merupakan toolkit yang powerfull sebagai analisis data diawal dan pengenalan struktur dataset dalam Python. Mengaktifkan Pandas sangat mudah didalam python. Cukup dengan melakukan import seperti dibawah ini
```
import pandas 
```
>Untuk mempermudah akses dan mempersingkat penamaan biasanya import pandas menggunakan inisial seperti dibawah ini:
```
Import pandas as pd
```
Setelah melakukan import Anda sudah bisa menggunakan pandas. Untuk lebih memahami lagi apa itu pandas silahkan membaca dokumentasi dari pandas agar mudah untuk melakukan eksplorasi dengan menggunakan pandas.





##1.1 Membaca file dengan menggunakan Pandas

Sebagai salah satu library untuk melakukan proses awal dari analisis data, pandas juga memiliki kemampuan untuk membaca berbagai macam jenis file. Format yang bisa dibaca oleh pandas ada berbagai macam, antara lain .txt, .csv, .tsv, dan lainnya. Pandas tidak hanya bisa membaca file saja, namun juga bisa merubah data dari file menjadi bentuk dataframe yang akhirnya nanti bisa diakses, diagregasi dan diolah. 
>Dataframe adalah sebuah tabel yang terdiri dari kolom dan baris dengan banyak tipe data di dalamnya 


In [0]:
import pandas as pd

Setelah sebelumnya kita mengenal library Pandas, sekarang kita akan belajar menggunakan library tersebut. Pandas memiliki fungsi untuk mengubah data dari berbagai jenis menjadi dataframe seperti yang telah dibahas di submodul sebelumnya. Contohnya, untuk mengubah data CSV menjadi dataframe, kita hanya cukup memanggil fungsi pandas.read_csv() dan memasukkan nama berkas pada parameternya. 

In [0]:
dataset1 = pd.read_csv('iris_train.csv') #mengubah data CSV menjadi dataframe

print(dataset1) #Mencetak Data


     sepal_length  sepal_width  petal_length  petal_width          species
0             5.1          3.4           1.5          0.2      Iris-setosa
1             6.3          2.5           5.0          1.9   Iris-virginica
2             5.9          3.0           4.2          1.5  Iris-versicolor
3             5.7          2.6           3.5          1.0  Iris-versicolor
4             7.7          2.8           6.7          2.0   Iris-virginica
..            ...          ...           ...          ...              ...
100           5.4          3.4           1.7          0.2      Iris-setosa
101           5.6          2.5           3.9          1.1  Iris-versicolor
102           6.4          2.9           4.3          1.3  Iris-versicolor
103           7.9          3.8           6.4          2.0   Iris-virginica
104           5.4          3.0           4.5          1.5  Iris-versicolor

[105 rows x 5 columns]


###Deskripsi Tabel Hasil Iris

![alt text](https://drive.google.com/uc?id=1AmdiDaujWpMDiZeFBLx5v6Hvk3uLWgQg)

## 1.2 Membaca file dengan menggunakan head()

Pada suatu kasus, data yang kita baca cukup banyak atau loading yang lama. Untuk memastikan data kita terbaca dengan baik dan bisa menampilkan data sebagian untuk ditampilkan secara benar, kita bisa memakai fungsi head().
fungsi head() tanpa parameter akan menghasilkan 5 baris data teratas yaitu data ke 0 - 4. Kita bisa gunakan parameter untuk menampilkan data teratas lebih dari 5 baris misal head(10) atau head(20) dan seterusnya

In [0]:
dataset1.head()

sepal_length  sepal_width  petal_length  petal_width          species
0           5.1          3.4           1.5          0.2      Iris-setosa
1           6.3          2.5           5.0          1.9   Iris-virginica
2           5.9          3.0           4.2          1.5  Iris-versicolor
3           5.7          2.6           3.5          1.0  Iris-versicolor
4           7.7          2.8           6.7          2.0   Iris-virginica

## 1.3 Melakukan akses data kolom

Dalam suatu analisis data ada kalanya kita hanya butuh melakukan akses beberapa data saja dan tidak perlu harus menampilkan semua data. Pada pandas kita bisa melakukan akses dalam berbagai kebutuhan. Mulai dari hanya akses kolom tertentu ataupun baris tertentu. Pada sesi kali ini kita akan mencoba untuk melakukan akses beberapa kolom tertentu pada suatu dataset.

In [0]:
dataset1.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

>*Note* :  Pada dataset ini ada 5 kolom termasuk class, dimana 4 kolom  merupakan data numerik dan 1 kolom merupakan data string

In [0]:
print(dataset1['species']) #data Label/Target species
print(dataset1['sepal_length']) #data attribut sepal_length

0          Iris-setosa
1       Iris-virginica
2      Iris-versicolor
3      Iris-versicolor
4       Iris-virginica
            ...       
100        Iris-setosa
101    Iris-versicolor
102    Iris-versicolor
103     Iris-virginica
104    Iris-versicolor
Name: species, Length: 105, dtype: object
0      5.1
1      6.3
2      5.9
3      5.7
4      7.7
      ... 
100    5.4
101    5.6
102    6.4
103    7.9
104    5.4
Name: sepal_length, Length: 105, dtype: float64


## 1.4 Melakukan akses data melalui baris

Selain melakukan akses data melalui kolom, dengan menggunakan pandas juga bisa melakukan akses dengan menggunakan baris. Berbeda dengan akses melalui kolom, fungsi untuk menampilkan data dari suatu baris adalah fungsi .iloc[i] dimana [i] menunjukan urutan baris yang akan ditampilkan yang dimana indexnya diawali dari 0. 



In [0]:
dataset1.iloc[7]

sepal_length                5.8
sepal_width                 2.7
petal_length                4.1
petal_width                   1
species         Iris-versicolor
Name: 7, dtype: object

## 1.5 Menampilkan suatu data dari baris dan kolom tertentu
Tidak hanya dengan menentukan dari kolom dan baris, dengan menggunakan pandas kita juga bisa memanggil suatu data dari suatu baris dan kolom tertentu dalam satu waktu. Perhatikan dan coba kode di bawah ini:

In [0]:
dataset1['sepal_length'].iloc[7] #kolom sepal_length baris ke 7

5.8

![alt text](https://drive.google.com/uc?id=1IIuq3aE9hdQa394sg0ap3opg7TV-EiZP)

## 1.6 Menampilkan data dalam range tertentu
Setelah menampilkan suatu kelompok data, bagaimana jika ingin menampilkan data dari baris ke 5 sampai ke 20 dari suatu dataset? Untuk mengantisipasi hal tersebut, pandas juga bisa menampilkan data dalam range tertentu, baik range untuk baris saja, kolom saja, dan range untuk baris dan kolom.

### Menampilkan range tertentu, kolom dan baris

In [0]:
dataset1['sepal_length'].iloc[5:21]

5     6.0
6     6.7
7     5.8
8     5.5
9     6.6
10    5.1
11    4.6
12    7.7
13    6.1
14    4.9
15    6.8
16    4.9
17    6.4
18    5.5
19    5.6
20    5.8
Name: sepal_length, dtype: float64

### Menampilkan range tertentu, baris saja

In [0]:
dataset1.iloc[5:21]

sepal_length  sepal_width  petal_length  petal_width          species
5            6.0          3.0           4.8          1.8   Iris-virginica
6            6.7          3.3           5.7          2.5   Iris-virginica
7            5.8          2.7           4.1          1.0  Iris-versicolor
8            5.5          2.3           4.0          1.3  Iris-versicolor
9            6.6          3.0           4.4          1.4  Iris-versicolor
10           5.1          3.7           1.5          0.4      Iris-setosa
11           4.6          3.6           1.0          0.2      Iris-setosa
12           7.7          2.6           6.9          2.3   Iris-virginica
13           6.1          2.8           4.0          1.3  Iris-versicolor
14           4.9          2.4           3.3          1.0  Iris-versicolor
15           6.8          3.0           5.5          2.1   Iris-virginica
16           4.9          3.1           1.5          0.1      Iris-setosa
17           6.4          2.7           5.3          1.9   Iris-virginica
18           5.5          2.4           3.8          1.1  Iris-versicolor
19           5.6          3.0           4.5          1.5  Iris-versicolor
20           5.8          2.8           5.1          2.4   Iris-virginica

## 1.7 Menampilkan informasi statistik dengan Numpy
Mengetahui informasi statistik pada suatu data sangat penting. Mulai dari distribusi data, nilai max atau min, hingga standar deviasi dari suatu dataset. Jika datanya berjumlah dibawah 10 mungkin masih dikerjakan secara manual. Namun, bayangkan jika datanya sudah mencapai ratusan bahkan ribuan. Tidak mungkin pastinya untuk dilakukan secara manual. Maka dari itu pentingnya fungsi describe() pada pandas. Fungsi describe() ini memungkinkan untuk mengetahui informasi statistik dari suatu dataset secara cepat.

In [0]:
dataset1.describe(include='all')


sepal_length  sepal_width  petal_length  petal_width          species
count     105.000000   105.000000    105.000000   105.000000              105
unique           NaN          NaN           NaN          NaN                3
top              NaN          NaN           NaN          NaN  Iris-versicolor
freq             NaN          NaN           NaN          NaN               41
mean        5.918095     3.042857      3.908571     1.249524              NaN
std         0.781613     0.437839      1.705935     0.736170              NaN
min         4.600000     2.000000      1.000000     0.100000              NaN
25%         5.200000     2.800000      1.700000     0.400000              NaN
50%         5.800000     3.000000      4.400000     1.400000              NaN
75%         6.400000     3.300000      5.100000     1.800000              NaN
max         7.900000     4.200000      6.900000     2.500000              NaN

>Note : Banyak nilai NaN yang tampil. Hal itu karena pada dataset ada format data string yang akhirnya memunculkan format NaN.
Untuk meminimalisir hal tersebut dan memfilter hanya data numerical saja, digunakan  
exclude=[‘O’], dimana fungsi itu akan mengabaikan data yang non-numerical untuk diproses. 

In [0]:
dataset1.describe(exclude='O') 
#Atau
dataset1.describe()

sepal_length  sepal_width  petal_length  petal_width
count    105.000000   105.000000    105.000000   105.000000
mean       5.918095     3.042857      3.908571     1.249524
std        0.781613     0.437839      1.705935     0.736170
min        4.600000     2.000000      1.000000     0.100000
25%        5.200000     2.800000      1.700000     0.400000
50%        5.800000     3.000000      4.400000     1.400000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.200000      6.900000     2.500000

# 2. Data Cleaning





Sebuah model Machine Learning tidak mampu untuk langsung / seketika mengolah data yang kita temukan dari berbagai sumber. Ada istilah Garbage In - Garbage Out yang berarti hasil dari machine learning akan buruk jika input yang Anda masukkan juga buruk. Tidak seluruh data yang kita dapat dari berbagai sumber siap untuk langsung diberikan ke sebuah model machine learning. Perolehan data memiliki  banyak kekurangan, sehingga perlu Anda olah terlebih dahulu.



1.   Konsistensi Format
Sebuah variabel mungkin tidak memiliki format yang konsisten seperti penulisan tanggal 10-Okt-2020 versus 10/10/20. Format jam yang berbeda seperti 17.10  versus 5.10 pm. Penulisan uang seperti 17000 versus Rp 17.000. Data dengan format berbeda tidak 

2.  Skala Data
Jika sebuah variabel memiliki jangka dari 1 sampai 100, pastikan tidak ada data yang lebih dari 100. Untuk data numerik, jika sebuah variabel merupakan bilangan positif, maka pastikan tidak ada bilangan negatif.



3.   Duplikasi data
Data yang memiliki duplikat akan mempengaruhi model machine learning, apalagi data yang duplikat memiliki jumlah yang besar. Untuk itu kita harus memastikan tidak ada data yang terduplikasi.

4. Skewness adalah kondisi di mana dataset cenderung memiliki distribusi data yang tidak seimbang. Skewness akan mempengaruhi data dengan menciptakan bias terhadap model. Apa itu bias? Sebuah model cenderung memprediksi sesuatu karena ia lebih sering mempelajari hal tersebut. Misalkan ada sebuah model untuk pengenalan buah di mana jumlah jeruk 92 buah dan apel 8 buah. Distribusi yang tidak imbang ini akan mengakibatkan model lebih cenderung memprediksi jeruk daripada apel.

5. Missing Value

   Penjelasan dibawah

     ![alt text](https://drive.google.com/uc?id=1HnxuJXuyHUqxSPI5WmdSVKNA5BcSs2pj)




## 5.   Missing Value
Missing value terjadi ketika data dari sebuah record tidak lengkap. Missing value sangat mempengaruhi performa model machine learning. Ada dua opsi untuk mengatasi missing value, yaitu menghilangkan data missing value atau mengganti nilai yang hilang dengan nilai lain, seperti rata-rata dari kolom tersebut atau nilai yang paling sering muncul.

>>Pada suatu dataset, ada kalanya data yang kita akan kita kelola tidak lengkap. Hal ini tentunya akan menyulitkan atau membuat hasil analisa tidak akurat. Penanggulangan akan data yang hilang ini biasa disebut **Handling Missing Value**. Penanganan dari nilai yang kosong ini banyak caranya. Sebagai seorang data science yang berhubungan dengan data yang real, solusi pertama yang benar-benar kita anjurkan untuk kasus seperti ini adalah melakukan trace kembali ke sumber data atau memerika ulang record. Terutama jika data itu berasal dari human record. Sangat disarankan untuk menelusuri kembali agar tidak terjadi kesalahan ketika sudah mencapai titik analisa. Selain solusi untuk melakukan penelusuran kembali ke sumberdata, pada ilmu data science juga ada beberapa metode yang bisa dijadikan solusi untuk menangani kasus ini.


### 5.1 Melakukan pengecekan untuk nilai NULL yang ada
Dengan menggunakan fungsi pandas, kita tidak perlu melihat satu persatu baris data untuk mengetahui apakah ada nilai kosong atau NULL/NAN pada suatu dataset. Bayangkan jika kita memilki 1000 baris data. Apakah kita harus melihat semua baris data tersebut? Tentu saja tidak. Maka dari itu di pandas disediakan fungsi untuk mengecek apakah ada data yang kosong.

```
.isnull().values.any()
```

In [0]:
dataset1.isnull().values.any()

False



>>Note : data yang digunakan merupakan data yang lengkap, maka dari itu output yang dihasilkan False



In [0]:
dataset2 = pd.read_csv("shopping_data_missingvalue.csv")

In [0]:
dataset2.isnull().values.any()

True

In [0]:
dataset2

CustomerID   Genre   Age  Annual Income (k$)  Spending Score (1-100)
0             1    Male  19.0                15.0                    39.0
1             2    Male   NaN                15.0                    81.0
2             3  Female  20.0                 NaN                     6.0
3             4  Female  23.0                16.0                    77.0
4             5  Female  31.0                17.0                     NaN
..          ...     ...   ...                 ...                     ...
195         196  Female  35.0               120.0                    79.0
196         197  Female  45.0               126.0                    28.0
197         198    Male  32.0               126.0                    74.0
198         199    Male  32.0               137.0                    18.0
199         200    Male  30.0               137.0                    83.0

[200 rows x 5 columns]

### Solusi mengisi Missing Value
Sebelum mengenal lebih banyak mengenai Solusi yang biasa ada pada kasus-kasus data science.

Skema yang biasanya sering dalam pengelolaan data.

![alt text](https://drive.google.com/uc?id=1aqxtfkCP0SgOBM2f69yxdhxsW1gJKIsq)

Dalam diagram diatas, perlu diketahui bahwa kasus kehilangan data bisa diatasi dengan berbagai cara. Bahkan, melakukan penghapusan data juga merupakan solusi yang bisa menjadi pilihan apabila jika dirasa mengisi nilai kosong akan memberikan pengaruh yang kurang bagus terhadap analisa, atau apabila pertimbangan data yang dihapus atau data yang hilang sedikit dan tidak memberikan terlalu banyak sumbangsih untuk analisa yang akan dilakukan. Penghapusan data bisa langsung pada baris data tersebut atau langsung satu kolom data. Pada solusi kedua yaitu menggunakan imputation (pengisian data yang kosong) bisa tergantung dari permasalahannya. Khusus untuk masalah yang berhubungan forecasting atau peramalan tergantung dari data yang ada (lebih lengkap bisa dilihat pada gambar). Khusus untuk general problem tergantung jenis datanya. Jika yang hilang data kategorikal atau bersifat string bisa menggunakna relasi antar kolom dengan Logistic Regression, jika numerical bisa menggunakan statistik sederhana dan linear regression. Pada sesi kali ini kita akan mencoba menangani data hilang dengan statistik sederhana, Mean dan Median.



### Mengisi dengan Mean
Salah satu metode yang bisa dikatakan sebagai solusi yang umum pada kasus general data science adalah mengisi data kosong dengan menggunakan mean dari masing-masing kolom. Pertama kita harus menentukan mean dari masing-masing kolom. Pada pandas terdapat fungsi mean() untuk menentukan nilai mean dari masing-masing kolom. Mean sendiri digunakan untuk 

>**data yang memiliki sedikit sifat outlier/noisy/anomali dalam sebaran datanya maupun isinya**






In [0]:
dataset2.mean() #mencari mean tiap kolom

CustomerID                100.500000
Age                        38.939698
Annual Income (k$)         61.005051
Spending Score (1-100)     50.489899
dtype: float64

In [0]:
dataset2_isi_mean=dataset2.fillna(dataset2.mean()) #mengisi missing value dengan data mean

In [0]:
dataset2_isi_mean #data sudah terisi 

CustomerID   Genre        Age  Annual Income (k$)  Spending Score (1-100)
0             1    Male  19.000000           15.000000               39.000000
1             2    Male  38.939698           15.000000               81.000000
2             3  Female  20.000000           61.005051                6.000000
3             4  Female  23.000000           16.000000               77.000000
4             5  Female  31.000000           17.000000               50.489899
..          ...     ...        ...                 ...                     ...
195         196  Female  35.000000          120.000000               79.000000
196         197  Female  45.000000          126.000000               28.000000
197         198    Male  32.000000          126.000000               74.000000
198         199    Male  32.000000          137.000000               18.000000
199         200    Male  30.000000          137.000000               83.000000

[200 rows x 5 columns]

### Mengisi dengan Median
Berbeda dengan mean pada sesi sebelumnya, median digunakan untuk data-data yang memiliki sifat outlier yang kuat. Kenapa median dipilih? Median merupakan nilai tengah yang artinya bukan hasil dari perhitungan yang melibatkan data outlier. Pada beberapa kasus, data outlier dianggap mengganggu dan sering dianggap noisy karena bisa mempengaruhi distribusi kelas dan mengganggu analisa pada klasterisasi (clustering).

>**median digunakan untuk data-data yang memiliki sifat outlier yang kuat**

In [0]:
dataset2_isi_median = dataset2.fillna(dataset2.median())
dataset2_isi_median

CustomerID   Genre   Age  Annual Income (k$)  Spending Score (1-100)
0             1    Male  19.0                15.0                    39.0
1             2    Male  36.0                15.0                    81.0
2             3  Female  20.0                62.0                     6.0
3             4  Female  23.0                16.0                    77.0
4             5  Female  31.0                17.0                    50.0
..          ...     ...   ...                 ...                     ...
195         196  Female  35.0               120.0                    79.0
196         197  Female  45.0               126.0                    28.0
197         198    Male  32.0               126.0                    74.0
198         199    Male  32.0               137.0                    18.0
199         200    Male  30.0               137.0                    83.0

[200 rows x 5 columns]

# 3. Data Preparation

Setelah dataset dibersihkan masih ada beberapa tahap yang perlu dilakukan agar dataset benar-benar siap untuk diproses oleh model machine learning. 

Umumnya beberapa model machine learning tidak dapat mengolah data kategorik, sehingga kita perlu melakukan konversi data kategorik menjadi data numerik. Data kategorik adalah data yang berupa kategori dan berjenis string. Contoh data kategorik adalah sebuah kolom pada dataset yang berisi jenis binatang seperti anjing, kucing, dan harimau. Contoh lain dari data kategorik adalah merek mobil seperti Ford, Honda, Toyota, dan BMW.

Banyak model machine learning seperti Regresi Linear dan Support Vector Machine (kedua model ini akan dibahas pada modul-modul selanjutnya) yang hanya menerima input numerik sehingga tidak bisa memproses data kategorik. Salah satu teknik untuk mengubah data kategorik menjadi data numerik adalah dengan menggunakan One Hot Encoding. One Hot Encoding mengubah data kategorik dengan membuat kolom baru untuk setiap kategori 

![alt text](https://drive.google.com/uc?id=120IgDDBfYGxBRAb195SIMaP6hd1TcqRq)


# 4. Data Preparation Opsional
Selain konversi data kategorik menjadi numerik, ada beberapa teknik lain dalam data preparation. Teknik yang akan dibahas antara lain membuang outlier, normalization, dan standardization.

## 4.1 Outlier Removal
Dalam statistik, outlier adalah sebuah nilai yang jauh berbeda dari kumpulan nilai lainnya dan dapat mengacaukan hasil dari sebuah analisis statistik. Outlier dapat disebabkan 

1. kesalahan dalam pengumpulan data 
2. atau nilai tersebut benar ada dan memang unik dari kumpulan nilai lainnya.

Apapun alasannya kemunculannya, Anda perlu  mengetahui cara untuk mengidentifikasi dan memproses outlier. Ini adalah bagian penting dalam persiapan data di dalam machine learning. Salah satu cara termudah untuk mengecek apakah terdapat outlier dalam data kita adalah dengan melakukan visualisasi.

![alt text](https://drive.google.com/uc?id=1f-S3JeHZl1VwP7ledcUxv58rtzL0ghVA)

Dapat dilihat dengan jelas bahwa terdapat satu sampel yang jauh berbeda dengan sampel-sampel lainnya. Setelah mengetahui bahwa di data kita terdapat outlier, kita dapat mencari lalu menghapus sampel tersebut dari dataset.



## 4.2 Normalization
### Sumber : Dicoding
Normalization adalah salah satu teknik yang dipakai dalam data preparation. Tujuan dari normalisasi adalah mengubah nilai-nilai dari sebuah fitur ke dalam skala yang sama. Normalization memungkinkan kenaikan performa dan stabilitas dari sebuah model machine learning.

![alt text](https://drive.google.com/uc?id=1xqjevGzHYdHj359GgD99huP-AQFDvFS2)

Contoh dari normalization adalah ketika kita memiliki dataset seperti di atas yang memiliki fitur umur dengan skala 23 sampai 45 tahun, dan fitur penghasilan dengan skala 4.000.000 sampai 35.000.000. Di sini kita melihat bahwa fitur penghasilan sekitar satu juta kali lebih besar dari fitur umur dan menunjukkan kedua fitur ini berada pada skala yang sangat jauh berbeda.

Ketika membangun model seperti regresi linear, fitur penghasilan akan sangat mempengaruhi prediksi dari model karena nilainya yang jauh lebih besar daripada umur. Walaupun, tidak berarti fitur tersebut jauh lebih penting dari fitur umur. 

Salah satu contoh dari normalization adalah min-max scaling di mana nilai-nilai dipetakan ke dalam skala 0 sampai 1. SKLearn menyediakan library untuk normalization

Pada Colab kita Import library MinMaxScaler dan masukkan data dari tabel sebelumnya.



```
from sklearn.preprocessing import MinMaxScaler
data = [[12000000, 33], [35000000, 45], [4000000, 23], [6500000, 26], [9000000, 29]]
```

Pada cell selanjutnya kita buat object scaler dan panggil fungsi fit() pada data. Fungsi fit adalah fungsi untuk menghitung nilai minimum dan maksimum pada tiap kolom.



```
scaler = MinMaxScaler()
scaler.fit(data)
```

Terakhir kita panggil fungsi transform yang akan mengaplikasikan scaler pada data.







In [0]:
from sklearn.preprocessing import MinMaxScaler
data = [[12000000, 33], [35000000, 45], [4000000, 23], [6500000, 26], [9000000, 29]]

scaler = MinMaxScaler()
scaler.fit(data)

print(scaler.transform(data))


[[0.25806452 0.45454545]
 [1.         1.        ]
 [0.         0.        ]
 [0.08064516 0.13636364]
 [0.16129032 0.27272727]]


Setiap nilai dari kolom gaji dan umur telah dipetakan pada skala yang sama seperti di bawah ini.

![alt text](https://drive.google.com/uc?id=1wa_HD7mHj5GBkJR1hWUdb9mam6BDj0e9)

Info Lengkap :https://scikit-learn.org/0.16/modules/generated/sklearn.preprocessing.MinMaxScaler.html

### Sumber : DQLab

Normalisasi Data


Terkadang pada beberapa kasus, 1 kolom dengan kolom yang lain memiliki skala yang berbeda. Seperti cuplikan gambar di bawah ini :

Antara Usia dan Masa Kerja masih memiliki range yang sama dalam skala puluhan. Namun, jika kolom Usia dan Masa Kerja dibandingkan dengan Gaji memiliki range nilai yang berbeda, dimana Usia dan Masa Kerja memiliki range puluhan dan Gaji mempunyai range nilai jutaan. Memang terlihat sederhana, namun hal ini bisa menjadi masalah besar dalam contoh kasus klasterisasi atau klasifikasi. Masuk pada kasus K-means yang sudah pernah dibahas sebelumnya. K-means merupakan algoritma klasterisasi (clustering) yang menggunakan perhitungan jarak dalam prosesnya. Sekarang coba bayangkan :

![alt text](https://drive.google.com/uc?id=1z_JnygGsZwubwc5gjVA1Jpa82mnXD9kn)

Jika tidak ada normalisasi, maka jelas perhitungan kmeans diatas akan tergantung pada Gaji. Kenapa? Karena gaji berdomain jutaan dan 2 kolom lainnya hanya berdomain puluhan.

>**Berapapun usia dan masa kerja seseorang tidak akan berpengaruh terhadap penilaian suatu perusahaan**. 

Perbedaan skala pada setiap kolom ini merupakan hal yang sangat wajar dan sering terjadi dan inilah pentingnya normalisasi. Normalisasi sangat penting, terutama untuk yang menggunakan perhitungan jarak dengan menggunakan metode apapun.



### Metode Normalisasi
Ada berbagai macam metode normalisasi, seperti 
1. MinMax, 
2. Zscore, 
3. Decimal Scaling, 
4. Sigmoid, dan 
5. Softmax. 

Pemakaiannya tergantung pada kebutuhan dari dataset dan jenis analisa yang dilakukan.

**1. MinMax**


>Metode Min-Max merupakan metode yang cukup bisa dibayangkan karena termasuk metode normalisasi yang bersifat linier dengan data aslinya. Namun, metode ini bisa menyebabkan out of bound pada beberapa kasus.

>![alt text](https://drive.google.com/uc?id=1J1knlg2ERpLKAi7xga4zMSnxu2qfh4ib)

>Kenapa bisa terjadi out of bound? Out of Bound terjadi apabila ada data baru masuk, dimana data tersebut melebihi nilai maksimal atau nilai minimal dari data yang sudah ada. Secara otomatis, perhitungan yang berlaku pada data yang sudah diperoleh tadi harus diulangi lagi semuanya dengan data baru yang masuk atau data baru yang mempunyai nilai maksimal/minimum yang melebihi tadi tidak bisa diproses. Karena kekurangan inilah MinMax tidak cocok untuk analisa real time / evolving system. Dimungkinkan dalam kasus-kasus terjadi kasus out of bound pada MinMax.

>MinMax sangat dianjurkan untuk kasus-kasus berbasis time frame analisis dan forecasting. Perhitungan dari metode ini cukup mengurangi data yang asli dengan nilai minimal dari fitur tersebut, kemudian hasil tersebut dikalikan dari hasil pengurangan nilai maximal yang baru dengan nilai minimal yang baru dan kemudian dibagi dengan nilai max dan min data di setiap fitur terakhir ditambah dengan nilai min yang baru.

**2. Z-Score**


>Zscore adalah metode yang sering digunakan dalam berbagai penelitian berbasis data mining atau data science. Z-score merupakan metode normalisasi yang berdasarkan mean (nilai rata-rata) dan standard deviation (deviasi standar) dari data. Kenapa Z-Score sangat populer? Selain tidak banyak variabel yang diset dalam perhitungannya. Z-Score sangat dinamis dalam melakukan perhitungan normalisasi. Kelemahan dari Z-Score adalah prosesnya akan terulang lagi jika ada data baru yang masuk. Selain itu elemen yang dibutuhkan untuk perhitungan Z-Score juga membutuhkan proses yang cukup lama baik standar deviation ataupun rata-rata dari setiap kolom.

![alt text](https://drive.google.com/uc?id=1dJYsU1X5gli1wVjiEhrDS69_y729YPYe)

![alt text](https://drive.google.com/uc?id=1DqwWEQgkqJ6itQrj3yj0_EI4vqb8hIxV)

## Praktek Normalisasi menggunakan Scikit Learn pada Python 

Scikit Learn merupakan library pada python yang digunakan untuk machine learning dan data science. Salah satu library yang selalu menjadi favorit dan komunitasnya sangat kuat. Scikit-learn sendiri tidak hanya untuk analytics saja, namun juga untuk pre-processing, feature selection, dan proses analysis lainnya. Melanjutkan dari sesi normalisasi data, mari kita praktekan kode di bawah ini :



In [0]:
import pandas as pd
import numpy as np
from sklearn import preprocessing

data = pd.read_csv("iris_train.csv")
array = data.values

In [0]:
print("sebelum dinormalisasi : ")
data.head(10)

sebelum dinormalisasi : 


sepal_length  sepal_width  petal_length  petal_width          species
0           5.1          3.4           1.5          0.2      Iris-setosa
1           6.3          2.5           5.0          1.9   Iris-virginica
2           5.9          3.0           4.2          1.5  Iris-versicolor
3           5.7          2.6           3.5          1.0  Iris-versicolor
4           7.7          2.8           6.7          2.0   Iris-virginica
5           6.0          3.0           4.8          1.8   Iris-virginica
6           6.7          3.3           5.7          2.5   Iris-virginica
7           5.8          2.7           4.1          1.0  Iris-versicolor
8           5.5          2.3           4.0          1.3  Iris-versicolor
9           6.6          3.0           4.4          1.4  Iris-versicolor

X merupakan matriks yang berisi fitur dataset yang akan digunakan dalam machine learning, baik untuk regresi, klasifikasi, pengklusteran, atau normalisasi
dan Y adalah matrik dari class/Labels

Pada kasus kita, X berisi fitur-fitur yang digunakan untuk dinormalisasi dengan teknik min-max scaler




![alt text](https://drive.google.com/uc?id=1AmdiDaujWpMDiZeFBLx5v6Hvk3uLWgQg)

>Pisahkan fitur dan class kedalam variabel X dan Y




In [0]:
X = array[:,:4] #memisahkan fitur dari dataset. 
Y = array[:,4:5]  #memisahkan class dari dataset

In [0]:
print("====Fitur :=====")
print(X)
print("--------------------------")
print()
print("====Class :=====")
print(Y)


====Fitur :=====
[[5.1 3.4 1.5 0.2]
 [6.3 2.5 5.0 1.9]
 [5.9 3.0 4.2 1.5]
 [5.7 2.6 3.5 1.0]
 [7.7 2.8 6.7 2.0]
 [6.0 3.0 4.8 1.8]
 [6.7 3.3 5.7 2.5]
 [5.8 2.7 4.1 1.0]
 [5.5 2.3 4.0 1.3]
 [6.6 3.0 4.4 1.4]
 [5.1 3.7 1.5 0.4]
 [4.6 3.6 1.0 0.2]
 [7.7 2.6 6.9 2.3]
 [6.1 2.8 4.0 1.3]
 [4.9 2.4 3.3 1.0]
 [6.8 3.0 5.5 2.1]
 [4.9 3.1 1.5 0.1]
 [6.4 2.7 5.3 1.9]
 [5.5 2.4 3.8 1.1]
 [5.6 3.0 4.5 1.5]
 [5.8 2.8 5.1 2.4]
 [5.8 2.7 5.1 1.9]
 [6.4 2.8 5.6 2.2]
 [6.7 3.0 5.2 2.3]
 [4.6 3.1 1.5 0.2]
 [6.9 3.1 5.4 2.1]
 [5.9 3.0 5.1 1.8]
 [6.2 2.8 4.8 1.8]
 [5.1 3.5 1.4 0.3]
 [5.8 2.7 3.9 1.2]
 [6.7 3.1 4.7 1.5]
 [6.3 2.5 4.9 1.5]
 [6.8 2.8 4.8 1.4]
 [6.2 2.2 4.5 1.5]
 [4.9 3.1 1.5 0.1]
 [5.2 4.1 1.5 0.1]
 [6.4 3.2 5.3 2.3]
 [5.0 2.3 3.3 1.0]
 [5.6 3.0 4.1 1.3]
 [6.3 3.4 5.6 2.4]
 [6.3 2.3 4.4 1.3]
 [6.3 2.8 5.1 1.5]
 [5.1 3.3 1.7 0.5]
 [6.3 2.9 5.6 1.8]
 [5.0 3.4 1.5 0.2]
 [7.2 3.0 5.8 1.6]
 [6.4 3.2 4.5 1.5]
 [5.0 2.0 3.5 1.0]
 [5.5 2.5 4.0 1.3]
 [5.1 3.8 1.6 0.2]
 [6.1 2.6 5.6 1.4]
 [6.2 2.9 4.3 

In [0]:
min_max_scaler = preprocessing.MinMaxScaler(feature_range=(0,1)) #inisialisasi normalisasi MinMax
data2 = min_max_scaler.fit_transform(X) #transformasi MinMax untuk fitur
dataset = pd.DataFrame({'sepal_length':data2[:,0],'sepal_width':data2[:,1],'petal_length Score':data2[:,2],'petal_width':data2[:,3],'species':array[:,4]})

print("dataset setelah dinormalisasi :")
print(dataset.head(10))

dataset setelah dinormalisasi :
   sepal_length  sepal_width  petal_length Score  petal_width          species
0      0.151515     0.636364            0.084746     0.041667      Iris-setosa
1      0.515152     0.227273            0.677966     0.750000   Iris-virginica
2      0.393939     0.454545            0.542373     0.583333  Iris-versicolor
3      0.333333     0.272727            0.423729     0.375000  Iris-versicolor
4      0.939394     0.363636            0.966102     0.791667   Iris-virginica
5      0.424242     0.454545            0.644068     0.708333   Iris-virginica
6      0.636364     0.590909            0.796610     1.000000   Iris-virginica
7      0.363636     0.318182            0.525424     0.375000  Iris-versicolor
8      0.272727     0.136364            0.508475     0.500000  Iris-versicolor
9      0.606061     0.454545            0.576271     0.541667  Iris-versicolor


## 4.3 Standardization
Standardization adalah proses untuk konversi nilai-nilai dari suatu fitur sehingga nilai-nilai tersebut memiliki skala yang sama. Z score adalah metode paling populer untuk standardisasi di mana setiap nilai pada sebuah atribut numerik akan dikurangi dengan rata-rata dan dibagi dengan standar deviasi dari seluruh nilai pada sebuah kolom atribut.

$ z = \frac{Nilai - Ratarata}{Standar Deviasi} $

Fungsi standarisasi sama dengan normalization. Keduanya berfungsi menyamakan skala nilai dari tiap atribut pada data. SKLearn menyediakan library untuk mengaplikasikan standard scaler pada data.

![alt text](https://drive.google.com/uc?id=1xqjevGzHYdHj359GgD99huP-AQFDvFS2)

Pada colab di cell pertama, kita akan mengimport library preprocessing dari scikit learn lalu membuat data dummy sesuai dengan tabel di atas.



```
from sklearn import preprocessing
data = [[12000000, 33], [35000000, 45], [4000000, 23], [6500000, 26], [9000000, 29]]
```

Selanjutnya kita buat object scaler dan panggil fungsi fit dari scaler pada data. Fungsi fit memiliki fungsi untuk menghitung rata-rata dan deviasi standar dari setiap kolom atribut untuk kemudian dipakai pada fungsi transform.



```
scaler = preprocessing.StandardScaler().fit(data)
```

Terakhir, kita panggil fungsi transform untuk mengaplikasikan standard scaler pada data. Untuk melihat hasil dari standard scaler kita tinggal memanggil objek scaler yang telah kita buat sebelumnya. Kodenya sebagai berikut.



```
data = scaler.transform(data)
data
```






In [0]:
from sklearn import preprocessing
data = [[12000000, 33], [35000000, 45], [4000000, 23], [6500000, 26], [9000000, 29]]

scaler = preprocessing.StandardScaler().fit(data)

data = scaler.transform(data)
data

array([[-0.11638732,  0.23521877],
       [ 1.94277296,  1.80334389],
       [-0.83261698, -1.07155217],
       [-0.60879521, -0.67952089],
       [-0.38497344, -0.28748961]])

https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.StandardScaler.html

# 5. Dataset

Dataset yang telah diproses kemudian siap kita proses dengan machine learning. Satu-satunya cara untuk mengetahui apakah model machine learning kita bagus atau tidak adalah dengan mengujinya pada kasus yang belum pernah ia hadapi. Kita bisa membuat sebuah model dan langsung mengujinya pada tahap produksi dan memonitor kualitasnya. Hal ini masuk akal, tapi jika ternyata model yang kita kembangkan bekerja dengan buruk, pelanggan dan klien kita akan komplain.

## 5.1 Training set dan Test set
Pilihan yang lebih baik adalah dengan membagi dataset menjadi 2 bagian yaitu data training dan data testing. Dengan demikian, kita bisa melakukan pelatihan model pada train set, dan mengujinya pada test set.

![alt text](https://drive.google.com/uc?id=1J2UyoLjFYZ8bilZRDZn98ZpOqBF9MD73)

Dengan menguji model terhadap data testing, kita dapat melihat kesalahan yang dibuat dan memperbaikinya sebelum mulai membawa model kita ke tahap produksi. Data testing diambil dengan proporsi tertentu 


> umumnya 20% dari keseluruhan data jika jumlah datanya kecil. Jika ukuran datanya sangat besar seperti 1 juta record, kita dapat mengambil sekitar 10 ribu data saja untuk testing alias sebesar 1% saja.




### Latihan SKLearn Train Test Split
Library Scikit Learn menyediakan fungsi untuk membagi dataset menjadi train set (data training) dan test set (data testing).

Library sklearn menyediakan dataset iris yakni sebuah dataset yang umum digunakan untuk masalah klasifikasi. Dataset ini memiliki jumlah 150 sampel. Untuk mendapatkan dataset, kita bisa menulis kode berikut pada cell baru.

In [0]:
import sklearn
from sklearn import datasets
from sklearn.model_selection import train_test_split

Dataset iris dari library sklearn belum dapat langsung dipakai oleh sebuah model ML. Sesuai dengan yang telah dibahas pada modul terdahulu, kita harus memisahkan antara atribut dan label pada dataset.

In [0]:
iris = datasets.load_iris()

In [0]:
x=iris.data #Attribut
y=iris.target #Label

Untuk membuat train set dan test set kita tinggal memanggil fungsi train_test_split. Train_test_split memiliki parameter x yaitu atribut dari dataset, y yaitu target dari dataset, dan test_size yaitu persentase dari test 

set dari dataset utuh. Train_test_split mengembalikan 4 nilai yaitu, atribut dari train set, atribut  dari test set, target dari train set, dan target dari test set.

In [0]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2) #  test size/train size = 20/80

Ketika kita print panjang dari x_test, kita bisa melihat bahwa panjang dari atribut test set adalah 30 sampel, sesuai dengan parameter yang kita masukkan pada fungsi train_test_split yaitu 0.2 atau 20% dari 150 sampel. Kode untuk print panjang dari x_test seperti di bawah ini

In [0]:
len(x_test)

30

Pada tahap ini dataset kita sudah bisa dipakai pada sebuah model machine learning.



# 6. Data Evaluation

Pada submodul sebelumnya kita telah belajar bagaimana menggunakan test set untuk mengevaluasi model sebelum masuk ke tahap produksi. 

Sekarang bayangkan ketika kita bertugas untuk mengembangkan sebuah proyek ML. Kita bimbang dalam memilih model yang akan dipakai, katakanlah dari 10 jenis model yang tersedia. Salah satu opsinya adalah dengan melatih kedua model tersebut lalu membandingkan tingkat erornya pada test set. Setelah membandingkan kedua model, Anda mendapati model regresi linier memiliki tingkat eror yang paling kecil katakanlah sebesar 5%. Anda lalu membawa model tersebut ke tahap produksi.

Kemudian ketika model diuji pada tahap produksi, tingkat eror ternyata sebesar 15%. Kenapa ini terjadi?. Masalah ini disebabkan karena kita mengukur tingkat eror berulang kali pada test set. Kita secara tidak sadar telah memilih model yang hanya bekerja dengan baik pada test set tersebut. Hal ini menyebabkan model tidak bekerja dengan baik ketika menemui data baru. Solusi paling umum dari masalah ini adalah dengan menggunakan validation set.



##Train test validation set
Validation set atau holdout validation adalah bagian dari train set yang dipakai untuk pengujian model pada tahap awal. Secara sederhana, kita menguji beberapa model dengan hyperparameter yang berbeda pada data training yang telah dikurangi data untuk validation. Lalu kita pilih model serta hyperparameter yang bekerja paling baik pada validation set. Setelah proses pengujian pada holdout validation, kita bisa melatih model menggunakan data training yang utuh (data training termasuk data validation) untuk mendapatkan model final. Terakhir kita mengevaluasi model final pada test set untuk melihat tingkat erornya.

![alt text](https://drive.google.com/uc?id=1-2_xD-sSFAWtFX45ayYiHoEDQX29DThQ)

Dalam menggunakan holdout validation, ada beberapa hal yang harus dipertimbangkan. Jika ukuran validation set-nya terlalu kecil, maka ada kemungkinan kita memilih model yang tidak optimal. Ketika ukuran validation set terlalu besar, maka sisa data pada train set lebih kecil dari data train set utuh di mana kondisi ini tidak ideal untuk membandingkan model yang berbeda pada data training yang lebih kecil. Solusi untuk masalah ini adalah dengan menggunakan **Cross Validation**

###Cross validation
K-Fold Cross Validation atau lebih sering disebut cross validation adalah salah satu teknik yang populer dipakai dalam evaluasi model ML. Pada cross validation, dataset dibagi sebanyak K lipatan. Pada setiap iterasi, setiap lipatan akan dipakai satu kali sebagai data uji dan lipatan sisanya dipakai sebagai data latih. Dengan menggunakan cross validation kita akan dapat hasil evaluasi yang lebih akurat karena model dievaluasi dengan seluruh data. Berikut adalah ilustrasi dari cross validation.

![alt text](https://drive.google.com/uc?id=1uvidr-8DO4cd4E2sicDa9WygBNWpvUiR)

## Latihan SKLearn Cross Validation Split

Library SKLearn menyediakan fungsi cross_val_score untuk menghitung hasil dari cross validation sebuah classifier bawaan sklearn. Pada coding practice kali ini kita akan menggunakan cross_validation_score pada classifier decision_tree. Dataset yang digunakan adalah dataset iris.

Pada Colab kita import semua library yang dibutuhkan di cell pertama. Dataset yang akan kita gunakan adalah dataset iris yang dipakai pada submodul sebelumnya.

In [0]:
import sklearn
from sklearn import datasets
from sklearn.model_selection import cross_val_score
from sklearn import  tree
iris = datasets.load_iris()

Kemudian kita bagi antara atribut dan label pada dataset.

In [0]:
x=iris.data
y=iris.target

Kita akan membuat model machine learning pertama kita yaitu decision tree, menggunakan library scikit learn. Model machine learning juga sering disebut sebagai classifier. Lebih lanjut, variabel clf adalah singkatan dari classifier.

In [0]:
clf = tree.DecisionTreeClassifier()

Setelah dataset dan model siap, kita bisa menggunakan cross validation untuk mengevaluasi performa dari model machine learning. Fungsi cross_val_score seperti di bawah menerima 4 parameter yaitu, ‘clf’ yang merupakan model machine learning, ‘X’ yang merupakan atribut dari dataset, ‘y’ yang merupakan label dari dataset, dan ‘cv’ yang merupakan jumlah fold yang akan dipakai pada cross validation.

In [0]:
score = cross_val_score(clf,x,y,cv=5)

Cross_val_score mengembalikan nilai berupa larik atau array yang terdiri dari akurasi pengujian setiap fold dari dataset. Untuk mencetak dan mengetahui hasilnya, tambahkan kode scores di bawah kode sebelumnya. Tampilannya seperti gambar di bawah ini.

In [91]:
score

array([0.96666667, 0.96666667, 0.9       , 0.93333333, 1.        ])

Elemen pertama dari larik menunjukkan nilai 0.96666 yang berarti ketika fold pertama dijadikan validation set dan fold lainnya dijadikan train set, hasil dari pengujian tersebut adalah akurasi sebesar 0.96666. 

Melihat akurasi dari seluruh pengujian fold yang memiliki nilai tinggi dan konsisten pada tiap fold, kita mendapatkan gambaran bahwa model kita memiliki performa yang sangat baik.

Secara umum jika hasil dari pengujian tiap fold pada cross validation memiliki nilai yang bervariasi dari 0.85 sampai 0.99, maka model tersebut dapat dikatakan baik.